File: articles/developmental-psychopathology-risk-resilience-and-adaptation/cpp/psychopathology_summary.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

std::vector<std::string> split_csv_line(const std::string& line) {
    std::vector<std::string> values;
    std::stringstream ss(line);
    std::string item;
    while (std::getline(ss, item, ',')) values.push_back(item);
    return values;
}

int find_column(const std::vector<std::string>& header, const std::string& name) {
    for (size_t i = 0; i < header.size(); ++i) {
        if (header[i] == name) return static_cast<int>(i);
    }
    return -1;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: psychopathology_summary data/developmental_psychopathology_panel.csv\n";
        return 1;
    }

    std::ifstream file(argv[1]);
    if (!file.is_open()) {
        std::cerr << "Unable to open CSV file\n";
        return 1;
    }

    std::string line;
    std::getline(file, line);
    auto header = split_csv_line(line);

    int score_col = find_column(header, "adaptation_score");
    int risk_col = find_column(header, "current_risk");
    int support_col = find_column(header, "current_support");

    long count = 0;
    double score_sum = 0.0, risk_sum = 0.0, support_sum = 0.0;

    while (std::getline(file, line)) {
        auto values = split_csv_line(line);
        if ((int)values.size() <= score_col || (int)values.size() <= risk_col || (int)values.size() <= support_col) continue;
        score_sum += std::stod(values[score_col]);
        risk_sum += std::stod(values[risk_col]);
        support_sum += std::stod(values[support_col]);
        count++;
    }

    std::cout << "Rows analyzed: " << count << "\n";
    std::cout << "Mean adaptation_score: " << score_sum / count << "\n";
    std::cout << "Mean current_risk: " << risk_sum / count << "\n";
    std::cout << "Mean current_support: " << support_sum / count << "\n";
    return 0;
}

File: articles/developmental-psychopathology-risk-resilience-and-adaptation/go/psychopathology_summary.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"
)

func col(header []string, name string) int {
	for i, h := range header {
		if h == name {
			return i
		}
	}
	return -1
}

func main() {
	if len(os.Args) < 2 {
		log.Fatal("Usage: go run psychopathology_summary.go data/developmental_psychopathology_panel.csv")
	}
	f, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	rows, err := csv.NewReader(f).ReadAll()
	if err != nil {
		log.Fatal(err)
	}
	if len(rows) < 2 {
		log.Fatal("CSV has no data rows")
	}

	header := rows[0]
	adaptIdx := col(header, "adaptation_score")
	riskIdx := col(header, "current_risk")
	supportIdx := col(header, "current_support")
	if adaptIdx < 0 || riskIdx < 0 || supportIdx < 0 {
		log.Fatal("Required columns missing")
	}

	count := 0
	adaptSum, riskSum, supportSum := 0.0, 0.0, 0.0

	for _, row := range rows[1:] {
		adapt, e1 := strconv.ParseFloat(row[adaptIdx], 64)
		risk, e2 := strconv.ParseFloat(row[riskIdx], 64)
		support, e3 := strconv.ParseFloat(row[supportIdx], 64)
		if e1 != nil || e2 != nil || e3 != nil {
			continue
		}
		count++
		adaptSum += adapt
		riskSum += risk
		supportSum += support
	}

	fmt.Printf("Rows analyzed: %d\n", count)
	fmt.Printf("Mean adaptation_score: %.4f\n", adaptSum/float64(count))
	fmt.Printf("Mean current_risk: %.4f\n", riskSum/float64(count))
	fmt.Printf("Mean current_support: %.4f\n", supportSum/float64(count))
}
